src/complete.jl

```diff
@@ -34,11 +35,64 @@ function n_complete(n::Int, D::Int)
     out
 end
 
-@generated function complete_polynomial!{N}(z::Matrix, d::Degree{N},
-                                            out::Matrix)
+#
+# Generating basis functions
+#
+@generated function complete_polynomial!{N,T,Ndim}(z::Array{T,Ndim}, d::Degree{N},
+                                                   out::Array{T,Ndim})
     complete_polynomial_impl!(z, d, out)
 end
 
+function complete_polynomial!{T, Ndim}(z::Array{T,Ndim}, d::Int, out::Array{T,Ndim})
+    complete_polynomial!(z, Degree{d}(), out)::Array{T,Ndim}
+end
+
+
+#
+# Vector versions for generating basis functions
+#
+function complete_polynomial_impl!{T,N}(z::Type{Vector{T}}, ::Type{Degree{N}},
+                                        out::Type{Vector{T}})
+    big_temp = Expr(:(=), Symbol("tmp_$(N+1)"), one(T))
+    quote
+        nvar = length(z)
+        if length(out) != (n_complete(nvar, $N))
+            error("z, out not compatible")
+        end
+
+        # reset first column to ones
+        out[1] = one($T)
+
+        ix = 1
+        $big_temp
+        @nloops($N, # number of loops
+                i,  # counter
+                d->((d == $N ? 1 : i_{d+1}) : nvar),  # ranges
+                d->(d == 1 ?
+                    (begin
+                         ix += 1
+                         out[ix] = tmp_{d+1}*z[i_d]
+                     end) :
+                    (begin
+                        ix += 1
+                        tmp_d = tmp_{d+1}*z[i_d]
+                        out[ix] = tmp_d
+                    end)),  # preexpr
+                Expr(:block, :nothing)  # bodyexpr
+                )
+        out
+    end
+end
```

Comment:
I agree that it is kind of a lot of code duplication -- Some of it was a playground for me to try and learn some stuff. A bit surprisingly, `complete_polynomial(reshape(z, 1, length(z)), 4)` actually does worse than `complete_polynomial(z', 4)` does. I use a slightly different way of constructing the points for vectors than for matrices (slight might even be an overstatement of the difference), but it does seem to continue being faster even as the number of dimensions and degree of polynomials are ramped up (we're talking about microseconds here...).

Thoughts? If specializing turns out to keep it faster then my vote would be to keep it separate for now, but I strongly suspect there are ways to close the gaps (I just haven't figured out how).